src/router.ts

```diff
@@ -0,0 +1,124 @@
+/**
+ * Agnostic router class
+ * Adapted from lukeed/trouter library:
+ * https://github.com/lukeed/trouter/blob/master/index.mjs
+ */
+import { parse } from "regexparam";
+import type {
+  FindResult,
+  FunctionLike,
+  HttpMethod,
+  Nextable,
+  RouteMatch,
+} from "./types.js";
+
+export type Route<H> = {
+  method: HttpMethod | "";
+  fns: H[];
+  isMiddle: boolean;
+} & (
+  | {
+      keys: string[] | false;
+      pattern: RegExp;
+    }
+  | { matchAll: true }
+);
+
+type RouteShortcutMethod<This, H extends FunctionLike> = (
+  route: RouteMatch | Nextable<H>,
+  ...fns: Nextable<H>[]
+) => This;
+
+export class Router<H extends FunctionLike> {
+  routes: Route<Nextable<H>>[];
+  constructor() {
+    this.routes = [];
+  }
+  public add(
+    method: HttpMethod | "",
+    route: RouteMatch | Nextable<H>,
+    ...fns: Nextable<H>[]
+  ): this {
+    if (typeof route === "function") {
+      fns.unshift(route);
+      route = "";
+    }
+    if (route === "")
+      this.routes.push({ matchAll: true, method, fns, isMiddle: false });
+    else {
+      const { keys, pattern } = parse(route);
+      this.routes.push({ keys, pattern, method, fns, isMiddle: false });
+    }
+    return this;
+  }
+  public all: RouteShortcutMethod<this, H> = this.add.bind(this, "");
+  public get: RouteShortcutMethod<this, H> = this.add.bind(this, "GET");
+  public head: RouteShortcutMethod<this, H> = this.add.bind(this, "HEAD");
+  public post: RouteShortcutMethod<this, H> = this.add.bind(this, "POST");
+  public put: RouteShortcutMethod<this, H> = this.add.bind(this, "PUT");
+  public patch: RouteShortcutMethod<this, H> = this.add.bind(this, "PATCH");
+  public delete: RouteShortcutMethod<this, H> = this.add.bind(this, "DELETE");
```

Comment:
Why is this using `this.add.bind(this)` over `() => this.add(this)` syntax? Just preference?